java/client/src/main/java/com/youtube/vitess/client/VTGateConn.java

```diff
@@ -0,0 +1,138 @@
+import com.youtube.vitess.grpc.proto.Vtgate.ExecuteKeyRangesResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.ExecuteKeyspaceIdsRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.ExecuteKeyspaceIdsResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.ExecuteShardsRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.ExecuteShardsResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.GetSrvKeyspaceRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.GetSrvKeyspaceResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteKeyRangesRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteKeyRangesResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteKeyspaceIdsRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteKeyspaceIdsResponse;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteShardsRequest;
+import com.youtube.vitess.grpc.proto.Vtgate.StreamExecuteShardsResponse;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+
+/**
+ * VTGateConn manages a VTGate connection.
+ *
+ * <p>Usage:
+ *
+ * <code>
+ *   RpcClient client = RpcClientFactory.create(
+ *     InetAddresses.forUriString("${VTGATE_ADDR}", Duration.millis(500)));
+ *   VTGateConn conn = VTGateConn.WithRpcClient(client);
+ *   Context ctx = Context.withDeadline(DateTime.now().plusMillis(500));
+ *
+ *   try {
+ *
+ *       CallerID callerId = CallerID.newBuilder().setPrincipal("username").build();
+ *       BindVariable bindVars = BindVariable.newBuilder()
+ *           .setType(Type.TYPE_INT)
+ *           .setValueInt(12345)
+ *           .build();
+ *       BoundQuery.Builder queryBuilder = BoundQuery.newBuilder()
+ *           .setSql(ByteString.copyFrom("INSERT INTO test_table VALUES(1, 2, 3)", Charsets.UTF_8));
+ *       queryBuilder.getMutableBindVariables().put("keyspaceid_01", bindVars);
+ *       BoundQuery query = queryBuilder.build();
+ *       ExecuteKeyspaceIdsRequest.newBuilder()
+ *           .setCallerId(callerId)
+ *           .setQuery(query)
+ *           .setKeyspace("my_keyspace")
+ *           .setKeyspaceIds(0, ByteString.copyFrom("keyspaceid_01", Charsets.UTF_8))
+ *           .setTabletType(TabletType.MASTER)
+ *           .build();
+ *
+ *       ExecuteKeyspaceIdsResponse response = conn.ExecuteKeyspaceIds(ctx, request);
+ *       if (response.hasError()) {
+ *         // handle error.
+ *       }
+ *       QueryResult result = response.getResult();
+ *       for (Row row : result.getRowsList()) {
+ *         // process each row.
+ *       }
+ *  } catch (VitessRpcException e) {
+ *     // ...
+ *   }
+ * </code>
+ * */
+public class VTGateConn implements Closeable {
+
+  private RpcClient client;
+
+  private VTGateConn(RpcClient client) { this.client = client; }
+
+  public static VTGateConn WithRpcClient(RpcClient client) {
+    return new VTGateConn(client);
+  }
+
+  public ExecuteShardsResponse ExecuteShard(Context ctx, ExecuteShardsRequest request)
```

Comment:
We need to define a VitessException for app-level errors and throw that. Like for integrity errors. It would have a code and a string.

But you can probably punt on that until Ammar has defined the error codes in a proto3 (and an error structure) and you can use them. Then in the gRPC implementation, you can plumb it in.

So for now, I would submit this in a form that works, and refactor it later (next week) for error codes.